### ISSUE TITLE
Add warning when comparing `sys.version_info` with old python

### ISSUE THREAD

**Issue by matusvalo** [PyCQA/pylint#7803]
### Current problem

Writing code which supports multiple python versions needs often do conditions based on python version - e.g.:
```python
if sys.version_info < (3, 5):
    # python 3.5 compatible code
else:
   # python 3.6 or later compatible code
```

Projects with longer history can have a lot of such statements. After dropping support of python 3.5 developer must grep all occurences of such conditions and clean up code that is dead.

### Desired solution

pylint should warn when conditions like `if sys.version_info < (3, 5)` occurs in code base and pylint parameter is `--py-version=3.6` or later. Ideally pylint should read `requires-python` configuration from `pyproject.toml` or `setup.cfg` but this is out of scope of this issue I suppose.

### Additional context

Ideally pylint should have a way to specify version of python supported for given line. This is common practise to leave comments in code which tends to be difficult to find or forgotten - see examples below:

* https://github.com/cython/cython/blob/efdf996dced075682d86b7e53a12b56b5ebfea38/tests/run/reraise.py#L34
* https://github.com/cython/cython/blob/f80d32583c4f6db5050ce48408415e7a55acaf02/Cython/Compiler/Lexicon.py#L55
* https://github.com/cython/cython/blob/efdf996dced075682d86b7e53a12b56b5ebfea38/tests/run/print_function.pyx#L2
* https://github.com/cython/cython/blob/1f65318a60660a2c4f07852e1341c1d48e19ea28/Cython/Compiler/Builtin.py#L176
* https://github.com/cython/cython/blob/66534408d3d8d6132e5b4e4061481661ba2d190e/Cython/Utils.py#L151
* https://github.com/cython/cython/blob/b4e91aa346dcea41f3e440113a0d92ea8139a23e/Cython/StringIOTree.py#L41

 Example how it can look like (python code from real codebase):

```python
try:
    from cStringIO import StringIO  # pylint: python <= 2.7
except ImportError:
    from io import StringIO
```

When one runs pylint with `--py-version=3.0` or greater it should raise a warning for line with pylint comment.

Some examples in Cython codebase:

**Comment by Pierre-Sassoulas** [comment]
> pylint should warn when conditions like if sys.version_info < (3, 5) occurs in the code base and pylint parameter is --py-version=3.6 or later. 

I like this, it's using an information we have in pylint, the check is going to be dead simple, and it's going to make py-version more useful, which is nice 👍 

> Ideally pylint should have a way to specify version of python supported for given line. 

Understanding a new pattern in comment though, it's something else entirely. This is a lot of work to implement and maintain and probably a performance hit too. Imo the benefit of it would be seen once the code base have actual comment inside them. But right now they don't. We also have a maintenance cost on the user side. But on the user side, if you have coverage and a continuous integration you're going to see that a piece of code is not tested anymore when you drop the interpreter so checking the coverage is probably going to be easier than adding comments in the code for pylint to parse in 2/3 years.

**Comment by matusvalo** [comment]
> Understanding a new pattern in comment though, it's something else entirely. 

Fair enough. Make sense what you wrote. I am more then fine having checker for `sys.version_info` and related attributes.

**Comment by DanielNoord** [comment]
Isn't this something `pyupgrade` fixes for us?

**Comment by matusvalo** [comment]
> Isn't this something `pyupgrade` fixes for us?

Not sure maybe just partially. But still even in astroid one can find such lines -e.g.:
https://github.com/PyCQA/astroid/blob/056d8e5fab7a167f73115d524ab92170b3ed5f9f/astroid/decorators.py#L22

**Comment by DanielNoord** [comment]
That check seems relevant though? There is support for 3.7 still right?

**Comment by matusvalo** [comment]
It is. My point is that the check should warn about all comparisons with python not supported by project anymore. So suppose you want to remove support of python 3.7 in astroid. You need to grep the code base and look for all `sys.version_info` usages and find this comparison and remove the code block because it is dead code after droping python 3.7. My proposal is to implement a warning in pylint.

**Comment by DanielNoord** [comment]
Yeah, but I think that check is already done by `pyupgrade`. If you change the supported version there, such `if ...` bodies get already removed

**Comment by matusvalo** [comment]
> If you change the supported version there, such if ... bodies get already removed

Hmm 🤔 I was not aware about that. I have not use pyupgrade... Hence I am leaving decision to you if you want to have this issue open or to close it.

**Comment by Pierre-Sassoulas** [comment]
I also think pyupgrade does fix it automatically. But I'm not against implementing checks that already exists in other tools as long as they are simple. Not everyone use all the tools that exists and we do warn about unused-import or variables that autoflake fix automatically. (Counterexample that happen often is that if mypy does it, it's probably going to be more painful to implement in pylint so that's the kind of check we sohuld not do.)

**Comment by DanielNoord** [comment]
> I also think pyupgrade does fix it automatically. But I'm not against implementing checks that already exists in other tools as long as they are simple. Not everyone use all the tools that exists and we do warn about unused-import or variables that autoflake fix automatically. (Counterexample that happen often is that if mypy does it, it's probably going to be more painful to implement in pylint so that's the kind of check we sohuld not do.)

Fully agree, just wanted to point this as out for people looking for this feature!

**Comment by SaharYaz** [comment]
Hi, I'd like to work on this issue. I've reproduced it on current main and have a local fix with tests.

**Comment by Pierre-Sassoulas** [comment]
Hey @SaharYaz I worked on this 2 weeks ago see https://github.com/pylint-dev/pylint/pull/11274. Not really ready or final but it was kinda lost on my fork so I opened a PR. It detected an old python 3.5 sys guard on pylint removed in https://github.com/pylint-dev/pylint/pull/11204. Don't hesitate to review and tell us what you think :)
